nomenclature/code.py

```diff
@@ -1,65 +1,145 @@
 
 
 class Code(BaseModel):
     """A simple class for a mapping of a "code" to its attributes"""
 
     name: str
+    description: Optional[str]
     attributes: Union[
-        Dict[str, Union[StrictStr, StrictInt, StrictFloat, StrictBool, List, None]],
+        Dict[
+            str,
+            Union[
+                StrictStr,
+                StrictInt,
+                StrictFloat,
+                StrictBool,
+                List,
+                None,
+                Dict[
+                    str,
+                    Union[StrictStr, StrictInt, StrictFloat, StrictBool, List, None],
+                ],
+            ],
+        ],
         List[StrictStr],
-    ]
+    ] = {}
 
     @classmethod
     def from_dict(cls, mapping):
         if isinstance(mapping, str):
-            return cls(name=mapping, attributes={})
+            return cls(name=mapping)
 
         if len(mapping) != 1:
             raise ValueError(f"Code is not a single name-attributes mapping: {mapping}")
 
-        return cls(name=list(mapping.keys())[0], attributes=list(mapping.values())[0])
+        # extract the name of the code
+        name = list(mapping.keys())[0]
+        # overwrite the mapping as just the code content
+        mapping = mapping[name]
+
+        # check if we have a "definition" attribute and map it to "description"
+        if "definition" in mapping:
+            if "description" not in mapping:
+                mapping["description"] = mapping["definition"]
+                del mapping["definition"]
+            else:
+                raise ValueError(
+                    f"Found both 'definition' and 'description' in {mapping}. "
+                    "Please use 'description'."
+                )
+
+        return cls(
+            name=name,
+            **{k: v for k, v in mapping.items() if k in cls.named_attributes()},
+            attributes={
+                k: v for k, v in mapping.items() if k not in cls.named_attributes()
+            },
+        )
 
     def set_attribute(self, key, value):
         self.attributes[key] = value
 
-
-class Tag(Code):
-    """A simple class for a mapping of a "{tag}" to "target codes" and attributes"""
-
-    attributes: List[
-        Dict[str, Union[str, Dict[str, Union[str, int, float, bool, List, None]], None]]
-    ]
-
-
-def replace_tags(code_list, tag, tag_dict):
-    """Replace tags in `code_list` by `tag_dict`"""
-
-    _code_list = []
-
-    for code in code_list:
-        if f"{{{tag}}}" in code.name:
-            _code_list.extend(_replace_tags(code, tag, tag_dict))
+    @classmethod
+    def named_attributes(cls) -> Set[str]:
+        return {a for a in cls.__dict__["__fields__"].keys() if a != "attributes"}
+
+    @property
+    def contains_tags(self) -> bool:
+        return re.search("{.*}", self.name) is not None
+
+    @property
+    def tags(self):
+        return re.findall("(?<={).*?(?=})", self.name)
+
+    def replace_tag(self, tag: str, target):
```

Comment:
target is always a Code, right? (Used to be a dict?)

Reply:
Yes target is always a `Code`. Since this is in the class definition of `Code` itself, just putting `target: Code` did not work. Luckily there's PEP484 to the rescue https://peps.python.org/pep-0484/#forward-references. Added the type hints now.